src/node_builtins.h

```diff
@@ -137,21 +135,27 @@ class NODE_EXTERN_PRIVATE BuiltinLoader {
       const v8::FunctionCallbackInfo<v8::Value>& args);
 
 #if defined(NODE_HAVE_I18N_SUPPORT)
-  static void AddExternalizedBuiltin(const char* id, const char* filename);
+  void AddExternalizedBuiltin(const char* id, const char* filename);
 #endif  // NODE_HAVE_I18N_SUPPORT
 
-  static BuiltinLoader instance_;
-  BuiltinCategories builtin_categories_;
+  // Used to synchronize access to builtin_categories.
+  // builtin_categories is marked mutable because it is only initialized
+  // as a cache, so that mutating it does not change any state.
+  Mutex builtin_categories_mutex_;
+  mutable std::optional<BuiltinCategories> builtin_categories_;
+
+  // Used to synchronize access to the code cache map
+  RwLock source_mutex_;
   BuiltinSourceMap source_;
-  BuiltinCodeCacheMap code_cache_;
-  UnionBytes config_;
 #if defined(NODE_HAVE_I18N_SUPPORT)
   std::list<std::unique_ptr<icu::UnicodeString>> externalized_source_bytes_;
 #endif  // NODE_HAVE_I18N_SUPPORT
 
-  // Used to synchronize access to the code cache map
-  Mutex code_cache_mutex_;
+  const UnionBytes config_;
 
+  // Used to synchronize access to the code cache map
+  RwLock code_cache_mutex_;
+  BuiltinCodeCacheMap code_cache_;
```

Comment:
Actually, only the `code_cache_` needs to be isolated, I think the rest like the config and the source code `UnionBytes` are shareable (and it's probably better to share them to save memory).

Reply:
Hmm, but we might also want to isolate these in case the embedder actually modifies the table...can we make it copy-on-write somehow? In the majority of the use cases, there isn't a need to modify the table, after all..

Reply:
Yeah, copy-on-write seems like the most reasonable thing to do in that case. I’ve added that in the most recent update here.